### PR TITLE
[CFE/circle-quantizer] Add --quantize_weights option

### PR DESCRIPTION
**Pull request by glistening**
It introduces --quantize_weights option in circle-quantizer.

ONE-DCO-1.0-Signed-off-by: Sanggyu Lee <sg5.lee@samsung.com>

Related: #11057

### CODE REVIEW

File: compiler/circle-quantizer/src/CircleQuantizer.cpp [hunk 1]
```diff
@@ -298,6 +308,19 @@ int entry(int argc, char **argv)
...
 
+  if (arser[qw])
+  {
+    auto values = arser.get<std::vector<std::string>>(qw);
+    if (values.size() != 3)
+    {
+      std::cerr << arser;
+      return 255;
+    }
+    // TODO: Add options and param
```
Comment (jinevening) [inline compiler/circle-quantizer/src/CircleQuantizer.cpp:319]: @glistening Are you going to support mixed-precision quantization too? Ex: some layers are quantized in uint8 and others are int16. If so, you may need to receive quantization configuration file too.

**Comment by jinevening** [comment]
Ah, I have another question. What is the dtype of quantized weights? Is it uint8 or int8? If it is different from our quantizer, we may need to extend interface.

File: compiler/circle-quantizer/src/CircleQuantizer.cpp (hunk 1 above)
Comment (glistening) [reply compiler/circle-quantizer/src/CircleQuantizer.cpp]: @jinevening  In short, no.

As I wrote in #11057, I am working this option for `onert` requirement.
`onert` does not have plan about mixed quantization for hybrid kernel.
(It is the reason I don't handle configuration option at all.)

Also, only one of {`channel` or `layer`} will be supported as `onert` supports 1 kind of kernels.

**Comment by glistening** [comment]
### How to use

```
$ ./build/debug/compiler/circle-quantizer/circle-quantizer \
--quantize_weights float32 uint8 channel in out
```

**Comment by glistening** [comment]
> Ah, I have another question. What is the dtype of quantized weights? Is it uint8 or int8? 

As I understand, `--quantize_weights` will have 3 parameters, exactly same to `--dequantize_quantize_weights`.

```cc
    .help("Quantize-dequantize weight values required action before quantization. "
          "Three arguments required: input_model_dtype(float32) "
          "output_model_dtype(uint8) granularity(layer, channel)");
```

I guess the configuration of

- output_model_dtype = uint8
- granularity = layer

is the kernel type that `onert` is (or will be) ready for.

( I know the default granularity is `channel` for weight.
But for `onert` execution, it is lower priority than layer-wise. )

**Comment by glistening** [comment]
I talked with @jinevening more.

- `circle-quantizer` does not support `int8`.
- It does not support symmetric ( -127 ~ 127 ) quantization.

I will check what kernel is ready more.

For `FullyConnected`, there are 3 version of kernels:

- CPU (C)
- NEON
- Ruy

I am thinking to support CPU (most portable) kernel, and we will cover `CONV`, `DCONV` more.

I will check all these cases.

cc @hseok-oh

**Comment by jinevening** [comment]
> I don't understand why we need to extend interface.interface.
(@jinevening Please let me know if I am missing something.)

As mentioned offline, I was curious if we will support int8 (this turns out to be true :) ), and if so, is it necessary to express symm/asymm in the interface.

**Comment by glistening** [comment]
@jinevening

1. I used wrong term in yesterday offline talk.
We don't introduce `symm/asymm`. It is about two modes in `symmetric` — `full range` vs `restricted range` (I've found circle quantization spec says this case as _narrow_ — e.g. INT16_narrow.)

2. I am not sure to add 4th arguments.
 
Let me check more. 

- triv2 quantization spec
- circle-quantizer implementation
- onert hybrid kernel implementation (fullyconnected, cov2d)
- tflite hybrid kernel implementation (dconv)

I will leave further comments in #11057, because it is PR. :)

**Comment by glistening** [comment]
Validation parameters added.
(core dumped is intended in debug build when runtime_exception is thrown.)

```
$ ./build/debug/compiler/circle-quantizer/circle-quantizer --quantize_weights \
a b c inception_v3.circle out
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unsupported input type. List of supported input type: float32
Aborted (core dumped)

$ ./build/debug/compiler/circle-quantizer/circle-quantizer --quantize_weights \
float32 b c inception_v3.circle out
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unsupported output type. List of supported output type: int8
Aborted (core dumped)

$ ./build/debug/compiler/circle-quantizer/circle-quantizer --quantize_weights \
float32 int8 c inception_v3.circle out
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unsupported granularity. List of supported granularity: layer
Aborted (core dumped)

$ ./build/debug/compiler/circle-quantizer/circle-quantizer --quantize_weights \
float32 int8 layer inception_v3.circle out
terminate called after throwing an instance of 'std::runtime_error'
  what():  QuantizeWeights is not supported yet
Aborted (core dumped)
```

**Comment by glistening** [comment]
#### layer-wise quantization is done

I think implementation is done based on QuantizeDequantizeWeights.

```
$ build/debug/compiler/luci/pass/luci_pass_test

...
[----------] 2 tests from QuantizeWeightsPassTest
[ RUN      ] QuantizeWeightsPassTest.name
[       OK ] QuantizeWeightsPassTest.name (0 ms)
[ RUN      ] QuantizeWeightsPassTest.name_ctx
[       OK ] QuantizeWeightsPassTest.name_ctx (0 ms)
[----------] 2 tests from QuantizeWeightsPassTest (0 ms total)

[----------] 3 tests from QuantizeWithMinMaxPassTest
[ RUN      ] QuantizeWithMinMaxPassTest.name
[       OK ] QuantizeWithMinMaxPassTest.name (0 ms)...
```

But the value test needs to be done. It will be the next step.

**Comment by glistening** [comment]
### templatize 

#### int8

```
$ ./build/debug/compiler/circle-quantizer/circle-quantizer \
--quantize_weights float32 int8 layer inception_v3.circle out.circle
```
```
$ build/debug/compiler/circledump/circledump out.circle

T(0:1) INT8 (32, 3, 3, 3) B(2) InceptionV3/InceptionV3/Conv2d_1a_3x3/Conv2D
    Quantization: min(-5.15335) max(5.15335) scale(0.000157273) zeropt(0) quantized_dimension(0)


T(0:4) INT8 (32, 3, 3, 32) B(5) InceptionV3/InceptionV3/Conv2d_2a_3x3/Conv2D
    Quantization: min(-1.96236) max(1.96236) scale(5.98884e-05) zeropt(0) quantized_dimension(0)
```
#### int16

```
$ ./build/debug/compiler/circle-quantizer/circle-quantizer \
--quantize_weights float32 int16 layer inception_v3.circle out.circle
```
```
$ build/debug/compiler/circledump/circledump out.circle

T(0:1) INT16 (32, 3, 3, 3) B(2) InceptionV3/InceptionV3/Conv2d_1a_3x3/Conv2D
    Quantization: min(-5.15335) max(5.15335) scale(0.000157273) zeropt(0) quantized_dimension(0)

T(0:2) FLOAT32 (32) B(3) InceptionV3/InceptionV3/Conv2d_1a_3x3/BatchNorm/FusedBatchNorm

T(0:3) FLOAT32 (1, 149, 149, 32) B(4) InceptionV3/InceptionV3/Conv2d_1a_3x3/Relu;InceptionV3/InceptionV3/Conv2d_1a_3x3/BatchNorm/FusedBatchNorm;InceptionV3/InceptionV3/Mixed_5b/Branch_3/Conv2d_0b_1x1/Conv2D;InceptionV3/InceptionV3/Conv2d_1a_3x3/Conv2D

T(0:4) INT16 (32, 3, 3, 32) B(5) InceptionV3/InceptionV3/Conv2d_2a_3x3/Conv2D
    Quantization: min(-1.96236) max(1.96236) scale(5.98884e-05) zeropt(0) quantized_dimension(0)
```

**Comment by hseok-oh** [comment]
>My understanding from https://github.com/Samsung/ONE/pull/11048 is that onert will interpret the weight values as `int8` for `uint8` type. (For backward compatibility of tflite's old quantization spec).

>Could you please cross-check the above?

On hybrid quantization, `uint8` weight is interpreted as symmetric `int8` internally.
When TOCO (old TensorFlow quantizer) try hybrid quantization, it quantize weight to `int8` type but mark as `uint8` type on TFLite model. And TFLite interpret quantized `uint8` weight as `int8` type if model uses hybrid quantization.
Circle follows TFLite's spec.

It is not true on full quantization. `uint8` full quantization (quzntize activation and weight) is `uint8`.

**Comment by hseok-oh** [comment]
FYI. `int8` type was introduced on `v1.12.1`.

https://www.github.com/tensorflow/tensorflow/commit/5fb82fc0184b199d302757bfb4f0a57921a8acf0

**Comment by glistening** [comment]
@jinevening, @hseok-oh 

Now, it is code for int8/int16 and per-channel quantization.

It generate weight-quantized circle for inception v3.

```
$ ./build/debug/compiler/circle-quantizer/circle-quantizer --quantize_weights float32 int8 channel inception_v3.circle s8.circle

$ ./build/debug/compiler/circle-quantizer/circle-quantizer --quantize_weights float32 int16 channel inception_v3.circle s16.circle

$ ls -lh
-rw-rw-r--   1 gyu  gyu   91M Nov  7  2022 inception_v3.circle
-rw-rw-r--   1 gyu  gyu   46M Aug 21 14:06 s16.circle
-rw-rw-r--   1 gyu  gyu   24M Aug 21 14:06 s8.circle
```

I confirmed with the size and the qparams and weight values from `netron-circle`.

### Next Steps in this PR

1. Add value test for `Conv2D` and `DConv2D` (In this PR)

2. Do not visit InstanceNorm, PReLU, ...

### Next Steps in next PRs

1. Templatize `sym_wquant_per_channel_s8` and `compute_sym_scale_s8` (In separate PR)

**Comment by glistening** [comment]
Value Test Done

```$ ctest --test-dir build/debug -R pota_wo* 
Internal ctest changing into directory: /ssd/z/ONE/build/debug
Test project /ssd/z/ONE/build/debug
    Start 107: pota_wo_quantization_test
1/1 Test #107: pota_wo_quantization_test ........   Passed    0.28 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.29 sec
```

Generated Files during Test
```
$ fd wo_q build/debug
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int16.wo_quantization.passed
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int16.wo_quantized.circle
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int16.wo_quantized.circle.h5
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int16_wo_quantization.log
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int8.wo_quantization.passed
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int8.wo_quantized.circle
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int8.wo_quantized.circle.h5
build/debug/compiler/pota-quantization-value-test/Conv2D_004.channel.int8_wo_quantization.log
```

For `int16`, I used the same data with quantization test.
For `int8`, I checked manually using spreadsheet. : )

I have to add `int8` support in `circle-tensordump`.

I am ready. I will split this PR and make PRs.

**Pull request closed without merging**